### PR TITLE
Tlwp 402 change links

### PR DESCRIPTION
**Pull request by adamesfa**

### CODE REVIEW

File: src/Sfa.Tl.Matching.Web/Controllers/OpportunityController.cs [hunk 1]
```diff
@@ -38,12 +38,15 @@ public async Task<IActionResult> CreateProvisionGap(CreateProvisionGapViewModel
...
-
             var createReferralViewModel = JsonConvert.DeserializeObject<CreateReferralViewModel>(viewModel);
 
             var dto = _mapper.Map<OpportunityDto>(createReferralViewModel);
-
-            var id = await _opportunityService.CreateOpportunity(dto);
+            
+            var id = createReferralViewModel.OpportunityId;
+            if (createReferralViewModel.OpportunityId > 0)
+                await _opportunityService.UpdateReferrals(dto);
```
Comment (hardik-desai-kcl) [inline src/Sfa.Tl.Matching.Web/Controllers/OpportunityController.cs]: you also need to update opportunity because route, postcode and Distance might have changed

File: src/Sfa.Tl.Matching.Web/Controllers/ProviderController.cs [hunk 2]
```diff
@@ -39,7 +39,14 @@ public IActionResult Start()
         [Route("find-providers", Name = "Providers_Get")]
         public IActionResult Index()
         {
-            return GetIndexViewAsync();
+            var viewModel = new SearchParametersViewModel
+            {
+                SelectedRouteId = null,
+                Postcode = null,
+                SearchRadius = SearchParametersViewModel.DefaultSearchRadius
+            };
```
Comment (hardik-desai-kcl) [inline src/Sfa.Tl.Matching.Web/Controllers/ProviderController.cs]: why this change?

File: src/Sfa.Tl.Matching.Web/Controllers/ProviderController.cs [hunk 3]
```diff
@@ -48,18 +55,19 @@ public async Task<IActionResult> Index(SearchParametersViewModel viewModel)
...
                 SelectedRouteId = viewModel.SelectedRouteId,
                 Postcode = viewModel.Postcode,
-                SearchRadius = SearchParametersViewModel.DefaultSearchRadius
+                SearchRadius = SearchParametersViewModel.DefaultSearchRadius,
+                OpportunityId = viewModel.OpportunityId
             });
         }
 
-        [Route("provider-results-within-{SearchRadius}-miles-of-{Postcode}-for-route-{SelectedRouteId}", Name = "ProviderResults_Get")]
+        [Route("provider-results-within-{SearchRadius}-miles-of-{Postcode}-for-route-{SelectedRouteId}-{OpportunityId}", Name = "ProviderResults_Get")]
```
Comment (hardik-desai-kcl) [inline src/Sfa.Tl.Matching.Web/Controllers/ProviderController.cs]: provider-results-for-opportunity-{OpportunityId}-within-{SearchRadius}-miles-of-{Postcode}-for-route-{SelectedRouteId}

File: src/Sfa.Tl.Matching.Web/Controllers/ProviderController.cs [hunk 4]
```diff
@@ -131,22 +137,40 @@ private async Task<SearchViewModel> GetSearchResultsAsync(SearchParametersViewMo
...
 
+            if (viewModel.OpportunityId == 0)
+                return resultsViewModel;
+
+            var selectedResultsViewModel = PopulateSelectedSearchResults(viewModel.OpportunityId, resultsViewModel);
+
+            return selectedResultsViewModel;
+        }
+
+        private SearchViewModel PopulateSelectedSearchResults(int opportunityId, SearchViewModel resultsViewModel)
```
Comment (hardik-desai-kcl) [inline src/Sfa.Tl.Matching.Web/Controllers/ProviderController.cs]: opportunityId is already set in SearchParametersViewModel.OpportunityId

Comment (hardik-desai-kcl) [inline src/Sfa.Tl.Matching.Web/Controllers/ProviderController.cs]: rename this function to SetProviderIsSelected

File: src/Sfa.Tl.Matching.Web/Views/Opportunity/_placementInformation.cshtml [hunk 5]
```diff
@@ -5,15 +5,38 @@
...
-            <td class="govuk-table__cell">@Model.RouteName</td>
+            <td class="govuk-table__cell">
+                @Model.RouteName
+            </td>
+            <td class="govuk-table__cell">
+                <a asp-route="ProviderResults_Get" 
+                   asp-route-opportunityId="@Model.OpportunityId" 
+                   asp-route-postcode="@Model.Postcode" 
+                   asp-route-selectedRouteId="@Model.RouteId" 
+                   class="govuk-link tl-float-right">Change<span class="govuk-visually-hidden"> the type of placement</span></a>
```
Comment (hardik-desai-kcl) [inline src/Sfa.Tl.Matching.Web/Views/Opportunity/_placementInformation.cshtml]: whats the point of having this hidden span?

File: src/Sfa.Tl.Matching.Application/Services/OpportunityService.cs [hunk 6]
```diff
@@ -37,12 +40,59 @@ public async Task<int> CreateOpportunity(OpportunityDto dto)
...
+            var dto = _mapper.Map<OpportunityDto>(opportunity);
+
+            return dto;
+        }
+
+        public OpportunityDto GetLatestCompletedOpportunity(Guid crmId)
+        {
+            var opportunities = _opportunityRepository.GetMany(o => o.EmployerCrmId == crmId,
+                o => o.ProvisionGap, 
+                o => o.Referral);
```
[the remark this answers is not in the document]
Comment (adamesfa) [reply src/Sfa.Tl.Matching.Application/Services/OpportunityService.cs]: Its being used to check whether the opportunities are valid. ie, in FilterValidOpportunities method. The opportunities are only looked at if are "complete".

File: src/Sfa.Tl.Matching.Web/Controllers/ProviderController.cs (hunk 2 above)
Comment (adamesfa) [reply src/Sfa.Tl.Matching.Web/Controllers/ProviderController.cs]: I made this refactoring because the number of parameters of the method was growing and too many primitive data types were being used.

File: src/Sfa.Tl.Matching.Application/Services/OpportunityService.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (hardik-desai-kcl) [reply src/Sfa.Tl.Matching.Application/Services/OpportunityService.cs]: I am questioning 
o => o.ProvisionGap, 
o => o.Referral);

this include is not required, and you can merge two where clauses into one at GetMany

File: src/Sfa.Tl.Matching.Web/Views/Opportunity/_placementInformation.cshtml (hunk 5 above)
Comment (adamesfa) [reply src/Sfa.Tl.Matching.Web/Views/Opportunity/_placementInformation.cshtml]: From Ash - "It's for screen readers
So when they're reading the content without visual clues it still makes sense
Otherwise tabbing through it'll just say "change" and not specify what it's doing"

File: src/Sfa.Tl.Matching.Web/Controllers/ProviderController.cs (hunk 4 above)
Comment (adamesfa) [reply src/Sfa.Tl.Matching.Web/Controllers/ProviderController.cs]: Correct. Ive removed the OpportunityId parameter

File: src/Sfa.Tl.Matching.Web/Controllers/ProviderController.cs (hunk 3 above)
Comment (adamesfa) [reply src/Sfa.Tl.Matching.Web/Controllers/ProviderController.cs]: Done

File: src/Sfa.Tl.Matching.Application/Services/OpportunityService.cs [hunk 7]
```diff
@@ -37,12 +40,59 @@ public async Task<int> CreateOpportunity(OpportunityDto dto)
...
+
+            var latestOpportunity = opportunities
```
[the remark this answers is not in the document]
Comment (adamesfa) [reply src/Sfa.Tl.Matching.Application/Services/OpportunityService.cs]: Done

File: src/Sfa.Tl.Matching.Models/Dto/EmployerDto.cs [hunk 8]
```diff
@@ -16,5 +16,6 @@ public class EmployerDto
         public string Postcode { get; set; }
         public string Owner { get; set; }
         public string CreatedBy { get; set; }
+        public string CompanyNameWithAka => $"{CompanyName} ({AlsoKnownAs})";
```
[the remark this answers is not in the document]
Comment (adamesfa) [reply src/Sfa.Tl.Matching.Models/Dto/EmployerDto.cs]: Done

File: src/Sfa.Tl.Matching.Web/Controllers/OpportunityController.cs (hunk 1 above)
Comment (adamesfa) [reply src/Sfa.Tl.Matching.Web/Controllers/OpportunityController.cs]: If any of these are updated, then its classed as a new opportunity and opportunity id would be 0 and go down the CreateOpportunity path. I believe this is the functionality that Luke wants.

**Comment by adamesfa** [comment]
Added comments.

File: src/Sfa.Tl.Matching.Application/Services/OpportunityService.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (adamesfa) [reply src/Sfa.Tl.Matching.Application/Services/OpportunityService.cs]: Removed. Will test.
